### ISSUE TITLE
Borrowing raw queries with mapper

### ISSUE THREAD

**Issue by Virgiel** [LouisGariepy/cornucopia#41]
Using borrowing raw query with a mapper lambda could give a very ergonomic way to deserialize struct with minimal allocation.

This code:

```sql
--! authors() #
SELECT name, country FROM Author
```

Would generate an additional function:
```rust
pub async fn authors_map<T: GenericClient, R>(
    client: &T,
    mut mapper: impl FnMut((&str, &str)) -> R,
) -> Result<impl Stream<Item = Result<R, Error>>, Error> {
    let stmt = client.prepare("SELECT name, country FROM Author;").await?;
    let stream = client
        .query_raw(&stmt, std::iter::empty::<u32>())
        .await?
        .map(move |res| {
            res.map(|res| {
                let return_value_1: &str = res.get(0);
                let return_value_2: &str = res.get(1);
                mapper((return_value_1, return_value_2))
            })
        });
    Ok(stream.into_stream())
}
```

Which can be used like that:
```rust
pub struct Person {
    name: String,
    country: String,
}

impl From<(&str, &str)> for Person {
    fn from((name, country): (&str, &str)) -> Self {
        Self {
            name: name.trim().to_owned(),
            country: country
                .split('|')
                .map(|s| s.trim())
                .nth(2)
                .unwrap_or_default()
                .to_owned(),
        }
    }
}

pub async fn get_authors<T: GenericClient>(client: &T) -> Result<Vec<Person>, Error> {
    authors_map(client, |raw| Person::from(raw)).await?.try_collect().await
}
```

I tried a function generic over type implementing `From` but it is a lifetime nightmare.

**Comment by LouisGariepy** [comment]
@Virgiel Thank you so much for your contribution. Looks promising, I'll give a shot at implementing it :+1: .

**Comment by LouisGariepy** [comment]
With some help, I was able to make my way through "lifetime hell". Take a look

```rust

pub struct AuthorsBorrowed<'a> {
    pub name: &'a str,
    pub country: &'a str,
}

pub struct Authors {
    pub name: String,
    pub country: String,
}

impl Authors {
    pub async fn map<T: GenericClient, R: for<'b> From<AuthorsBorrowed<'b>>>( //<-- Notice this `for` clause! That's how I solved it.
        client: &T,
    ) -> Result<impl Stream<Item = Result<R, Error>>, Error> {
        let stmt = client.prepare("SELECT name, country FROM Author;").await?;
        let stream = client
            .query_raw(&stmt, std::iter::empty::<u32>())
            .await?
            .map(|res| {
                res.map(|res| {
                    R::from(AuthorsBorrowed {
                        name: res.get(0),
                        country: res.get(1),
                    })
                })
            });
        Ok(stream.into_stream())
    }

    pub async fn query<T: GenericClient>(
        client: &T,
    ) -> Result<impl Stream<Item = Result<Authors, Error>>, Error> {
        let stmt = client.prepare("SELECT name, country FROM Author;").await?;
        let stream = client
            .query_raw(&stmt, std::iter::empty::<u32>())
            .await?
            .map(|res| {
                res.map(|res| Authors {
                    name: res.get(0),
                    country: res.get(1),
                })
            });
        Ok(stream.into_stream())
    }
}
```
And this is how you could use it.
```rust
pub struct CustomAuthors {
    pub name: String,
    pub country: String,
}

impl<'a> From<AuthorsBorrowed<'a>> for CustomAuthors {
    fn from(AuthorsBorrowed { name, country }: AuthorsBorrowed) -> Self {
        Self {
            name: name.trim().to_owned(),
            country: country
                .split('|')
                .map(|s| s.trim())
                .nth(2)
                .unwrap_or_default()
                .to_owned(),
        }
    }
}

pub async fn get_authors<'a, T: GenericClient>(client: &T) -> Result<Vec<Authors>, Error> {
    Authors::query(client).await?.try_collect().await
}

pub async fn get_authors_custom_map<T: GenericClient>(
    client: &T,
) -> Result<Vec<CustomAuthors>, Error> {
    Authors::map(client).await?.try_collect().await
}

```

We could do without the "owned" version of the struct and without the `query` method, but I thought they'd be nice to have for users that don't need any mapping at all.

There might be ways to make this even more ergonomic though, please let me know what you think :+1:

**Comment by Virgiel** [comment]
This is lifetime dark magic 🤯 even better than I hoped !

We could do this:
```rs
pub struct AuthorsBorrowed<'a> {
    pub name: &'a str,
    pub country: &'a str,
}

pub struct Authors {
    pub name: String,
    pub country: String,
}

impl<'a> From<AuthorsBorrowed<'a>> for Authors {
    fn from(AuthorsBorrowed { name, country }: AuthorsBorrowed<'a>) -> Self {
        Self {
            name: name.to_owned(),
            country: country.to_owned(),
        }
    }
}

pub async fn authors<T: GenericClient, R: for<'b> From<AuthorsBorrowed<'b>>>(
    client: &T,
) -> Result<impl Stream<Item = Result<R, Error>>, Error> {
    let stmt = client.prepare("SELECT name, country FROM Author;").await?;
    let stream = client
        .query_raw(&stmt, std::iter::empty::<u32>())
        .await?
        .map(|res| {
            res.map(|res| {
                R::from(AuthorsBorrowed {
                    name: res.get(0),
                    country: res.get(1),
                })
            })
        });
    Ok(stream.into_stream())
}

pub struct CustomAuthors {
    pub name: String,
    pub country: String,
}

impl<'a> From<AuthorsBorrowed<'a>> for CustomAuthors {
    fn from(AuthorsBorrowed { name, country }: AuthorsBorrowed) -> Self {
        Self {
            name: name.trim().to_owned(),
            country: country
                .split('|')
                .map(|s| s.trim())
                .nth(2)
                .unwrap_or_default()
                .to_owned(),
        }
    }
}

pub async fn get_authors<'a, T: GenericClient>(client: &T) -> Result<Vec<Authors>, Error> {
    authors(client).await?.try_collect().await
}

pub async fn get_authors_custom_map<T: GenericClient>(
    client: &T,
) -> Result<Vec<CustomAuthors>, Error> {
    authors(client).await?.try_collect().await
}
```

It might be complicated for users to guess that an owned variant exists. But if this is well documented this would be very ergonomic.

If merged with another of my [proposal](https://github.com/LouisGariepy/cornucopia/discussions/42), we could hide the borrowed type behind the convenient API functions.

**Comment by Virgiel** [comment]
Another question I had is if it is possible to also handle `TryFrom` trait. But in those cases, this means the database content is incompatible with the structure we try to deserialize. Therefore, I am not sure this would be useful.

**Comment by LouisGariepy** [comment]
As mentioned in https://github.com/LouisGariepy/cornucopia/discussions/42, I'll give a shot at implementing your suggestions.

> possible to also handle TryFrom trait...
In my view, the mapper should be infallible (or handle the failures internally). But I'm not against it per se. If someone has a convincing use case, I'm definitely open to add support for it.

**Comment by LouisGariepy** [comment]
After much experimentation, it doesn't seem possible to merge the `From` mapping proposal and the `Query` client code proposal, due to borrow checker limitations. Its a bit hard for me to explain, but I discussed it in length over at the Rust discord and I can provide details if needed.

To be clear, this is not really problematic, it just makes the generated code a bit more verbose, but I don't think there are any downsides apart from that.

**Comment by Virgiel** [comment]
The verbosity of the generated code is not a huge problem. We can always try to reduce it later if it becomes problematic.
If you have a discord link to your posts, I would be interested in reading the discussion, if it's too complicated don't bother.

**Comment by LouisGariepy** [comment]
Sure thing :) I'll make a loose transcript here though, just for the record if this becomes relevant.

So, for query to be `Query` to be usable by all queries, it (or its methods) has to be generic over types representing say `Authors` and `AuthorsBorrowed`. So, you end up with code like this

```rust
async fn one<A, B>(/*whatever*/) -> Result<B, Error>
where
    A: for<'b> From<&'b Row>,
    B: From<A>,
{
    let row = client.query_one(/*whatever*/).await?;
    Ok(B::from(A::from(&row)))
}
```

Unfortunately, there are no way to make this work because `Implementation of From is not general enough. From<&'0 Row> would have to be implemented for the type A<'_> for any lifetime, but From<&'1 Row> is actually implemented for the type A<'1>, for some specific lifetime '1`. There is no way to name that lifetime :/

The other way to go about it is 
```rust
async fn one<'a, A,B>() -> Result<B, Error>
where
    A: From<&'a Row> + 'a,
    B: From<A>,
{
    let row = client.query_one(/*whatever*/).await?;
    Ok(B::from(A::from(&row)))
}
```
But now Rust complains that we're `trying to return a borrow to a temporary `, which is not the case as far as I understand since B is owned, but again this seems to be a borrow-checker limitation?

**Comment by LouisGariepy** [comment]
@Virgiel To be clear, what I'm currently aiming for would look something like this 
```rust
pub struct AuthorsBorrowed<'a> {
    pub name: &'a str,
    pub country: &'a str,
}

pub struct Authors {
    pub name: String,
    pub country: String,
}

impl<'a> From<AuthorsBorrowed<'a>> for Authors {
    fn from(AuthorsBorrowed { name, country }: AuthorsBorrowed<'a>) -> Self {
        Self {
            name: name.to_owned(),
            country: country.to_owned(),
        }
    }
}

impl Authors {
    pub async fn stmt<C: GenericClient>(client: &C) -> Result<Statement, Error> {
        client.prepare("SELECT name, country FROM Author;").await
    }

    pub async fn stream<C: GenericClient, R: for<'b> From<AuthorsBorrowed<'b>>>(
        client: &C,
    ) -> Result<impl Stream<Item = Result<R, Error>>, Error> {
        let stmt = Self::stmt::<C>(&client).await?;
        let stream = client
            .query_raw(&stmt, std::iter::empty::<u32>())
            .await?
            .map(|res| {
                res.map(|res| {
                    R::from(AuthorsBorrowed {
                        name: res.get(0),
                        country: res.get(1),
                    })
                })
            });
        Ok(stream.into_stream())
    }

    pub async fn vec<C: GenericClient>(client: &C) -> Result<Vec<Self>, Error> {
        Self::stream(client).await?.try_collect().await
    }

    pub async fn one<C: GenericClient>(client: &C) -> Result<Self, Error> {
        let stmt = Self::stmt::<C>(&client).await?;
        let row = client.query_one(&stmt, &[]).await?;
        Ok(Self {
            name: row.get(0),
            country: row.get(1),
        })
    }

    pub async fn opt<C: GenericClient>(client: &C) -> Result<Option<Self>, Error> {
        let stmt = Self::stmt::<C>(&client).await?;
        Ok(client.query_opt(&stmt, &[]).await?.map(|row| Self {
            name: row.get(0),
            country: row.get(1),
        }))
    }
}
```

What do you think?

**Comment by Virgiel** [comment]
Unfortunately you are right. I don't see how we can reuse a query structure. However, I prefer to keep a query structure separate from the return type for compatibility with other of my proposals.

Let's start by agreeing that we will generate borrowed and owned versions only if there is a borrowable type in the rows. As soon as we use a generic function, users lose the inference and have to explicitly specify the return type (unfortunately, default value for type parameters does not work for functions). Therefore, I see no advantage in allowing `From<Owned>` instead of `<'a> From<Borrowed<'a>`. It may seem less ergonomic, but if users are going to use their custom types, we shouldn't let them implement an unoptimized version.

My proposal is the following. This code:

```sql
--! authors
SELECT name WHERE city = :city FROM Author;
```

Generate those types:

```rs
pub struct AuthorBorrowed<'a> {
    name: &'a str,
}

pub struct Author {
    name: String,
}

impl<'a> From<AuthorBorrowed<'a>> for Author {
    fn from(it: AuthorBorrowed<'a>) -> Self {
        Author {
            name: it.name.to_owned(),
        }
    }
}
```

And this query code:

```rs
pub fn authors<'a, 'b, C: GenericClient>(client: &'a C, city: &'a &str) -> AuthorQuery<'a, C> {
    AuthorQuery {
        client,
        params: [city],
    }
}

pub struct AuthorQuery<'a, C: GenericClient> {
    client: &'a C,
    params: [&'a (dyn tokio_postgres::types::ToSql + Sync); 1],
}

impl<'a, C> AuthorQuery<'a, C>
where
    C: GenericClient,
{
    pub fn mapper<'b, R: From<AuthorBorrowed<'b>>>(r: &'b Row) -> R {
        let borrow = AuthorBorrowed { name: r.get(0) };
        R::from(borrow)
    }

    pub async fn stmt(&self) -> Result<Statement, Error> {
        self.client.prepare("SELECT name FROM Author;").await
    }

    pub async fn one<T: for<'b> From<AuthorBorrowed<'b>>>(self) -> Result<T, Error> {
        let stmt = self.stmt().await?;
        let row = self.client.query_one(&stmt, &self.params).await?;
        Ok(Self::mapper(&row))
    }

    pub async fn list<T: for<'b> From<AuthorBorrowed<'b>>>(self) -> Result<Vec<T>, Error> {
        self.raw().await?.try_collect().await
    }

    pub async fn opt<T: for<'b> From<AuthorBorrowed<'b>>>(self) -> Result<Option<T>, Error> {
        let stmt = self.stmt().await?;
        Ok(self
            .client
            .query_opt(&stmt, &self.params)
            .await?
            .map(|r| Self::mapper(&r)))
    }

    pub async fn raw<T: for<'b> From<AuthorBorrowed<'b>>>(
        self,
    ) -> Result<impl Stream<Item = Result<T, Error>>, Error> {
        let stmt = self.stmt().await?;
        let stream = self
            .client
            .query_raw(&stmt, slice_iter(&self.params))
            .await?
            .map(move |res| res.map(|r| Self::mapper(&r)));
        Ok(stream.into_stream())
    }
}

// https://github.com/sfackler/rust-postgres/blob/765395f288861209a644c621bf72172acd482515/tokio-postgres/src/lib.rs
fn slice_iter<'a>(
    s: &'a [&'a (dyn ToSql + Sync)],
) -> impl ExactSizeIterator<Item = &'a dyn ToSql> + 'a {
    s.iter().map(|s| *s as _)
}
```
Used like this:

```rs
pub async fn usage<T: GenericClient>(client: &T) -> Result<(), Error> {
    let author: Author = authors(client, &"Paris").one().await?;
    let user_author: UserAuthor = authors(client, &"Paris").one().await?;
    let author = authors(client, &"Paris").one::<Author>().await?;
    let user_author = authors(client, &"Paris").one::<UserAuthor>().await?;
    Ok(())
}
```

If you think loss of inference is a problem, we could have a version of the ergonomic function that always returns the owned version type and others that are generic:

```rs
impl<'a, C> AuthorQuery<'a, C>
where
    C: GenericClient,
{
    pub async fn one(self) -> Result<Author, Error> {
        self.one_custom().await
    }

    pub async fn list(self) -> Result<Vec<Author>, Error> {
        self.list_custom().await
    }

    pub async fn opt(self) -> Result<Option<Author>, Error> {
        self.opt_custom().await
    }

    pub async fn one_custom<T: for<'b> From<AuthorBorrowed<'b>>>(self) -> Result<T, Error> {
        let stmt = self.stmt().await?;
        let row = self.client.query_one(&stmt, &self.params).await?;
        Ok(Self::mapper(&row))
    }

    pub async fn list_custom<T: for<'b> From<AuthorBorrowed<'b>>>(self) -> Result<Vec<T>, Error> {
        self.raw().await?.try_collect().await
    }

    pub async fn opt_custom<T: for<'b> From<AuthorBorrowed<'b>>>(self) -> Result<Option<T>, Error> {
        let stmt = self.stmt().await?;
        Ok(self
            .client
            .query_opt(&stmt, &self.params)
            .await?
            .map(|r| Self::mapper(&r)))
    }

    pub async fn raw<T: for<'b> From<AuthorBorrowed<'b>>>(
        self,
    ) -> Result<impl Stream<Item = Result<T, Error>>, Error> {
        let stmt = self.stmt().await?;
        let stream = self
            .client
            .query_raw(&stmt, slice_iter(&self.params))
            .await?
            .map(move |res| res.map(|r| Self::mapper(&r)));
        Ok(stream.into_stream())
    }
}
```

```rs
pub async fn usage<T: GenericClient>(client: &T) -> Result<(), Error> {
    let infered = authors(client, &"Paris").one().await?;
    let author: Author = authors(client, &"Paris").one_custom().await?;
    let user_author: UserAuthor = authors(client, &"Paris").one_custom().await?;
    let author = authors(client, &"Paris").one_custom::<Author>().await?;
    let user_author = authors(client, &"Paris").one_custom::<UserAuthor>().await?;
    Ok(())
}
```

**Comment by Virgiel** [comment]
The only specific things are the query function, the mapper function, the query string and the borrowed type. We don't rely on code reuse anymore but we keep the codegen relatively simple. What do you think about it?

**Comment by LouisGariepy** [comment]
> I prefer to keep a query structure separate from the return type for compatibility with other of my proposals.
No objection with that :+1: 

> Let's start by agreeing that we will generate borrowed and owned versions only if there is a borrowable type in the rows. Could you clarify please, I'm not sure I'm following :smile:. What I've done as of now is to not generate the structs if there are no return columns, for example, but I'm not sure if that's what you're referring to.

> If you think loss of inference is a problem
I was mostly concerned with saving a few keystrokes for the most common use cases, but losing it is fine as far as I'm concerned.

I'm pretty satisfied with this design and I'll start implementing it if we're satisfied with it :+1:

**Comment by Virgiel** [comment]
> Could you clarify please, I'm not sure I'm following 😄. What I've done as of now is to not generate the structs if there are no return columns, for example, but I'm not sure if that's what you're referring to.

I do not want useless borrowed variant:

```sql
--! authors
SELECT name WHERE city = :city FROM Author;

--! authors_age
SELECT age WHERE city = :city FROM Author;
```

```rs
// Borrowed version because &str can be borrowed
pub struct AuthorsBorrowed<'a> {
    name: &'a str,
}

pub struct Authors {
    name: String,
}

impl<'a> From<AuthorsBorrowed<'a>> for Authors {
    fn from(it: AuthorsBorrowed<'a>) -> Self {
        Authors {
            name: it.name.to_owned(),
        }
    }
}

// Only owned because there is nothing to borrow
pub struct AuthorsAge {
    age: i64,
}
```

But thinking about it, handling these cases would increase the complexity of the code, so we should look at how hard it is to implement.

**Comment by Virgiel** [comment]
> I was mostly concerned with saving a few keystrokes for the most common use cases, but losing it is fine as far as I'm concerned. It wouldn't be breaking either to add "one_owned" etc later on.

It is less user-friendly because the user has to know that there is a structure with a name similar to the function to import and use. If Rust supported defining a [default type](https://rust-lang.github.io/rfcs/0213-defaulted-type-params.html) we would not have this problem.

**Comment by LouisGariepy** [comment]
> I do not want useless borrowed variant:

Ok now I see! Yeah this could be a nice improvement. I won't block this specific issue on it though, I think it'd  be better to move it to its own issue.

If the type is basic (like `INT`/`i32`) this is not too hard to implement. Its harder when you consider composites/domains though because you have to recursively check their inner types. Still, this seems plausibly doable.

Someone has also asked something very similar but for the param struct (i.e. not have to pass `Copy` types as references). I guess implementing one would get use the other for free.

**Comment by LouisGariepy** [comment]
> > I was mostly concerned with saving a few keystrokes for the most common use cases, but losing it is fine as far as I'm concerned. It wouldn't be breaking either to add "one_owned" etc later on.
> 
> It is less user-friendly because the user has to know that there is a structure with a name similar to the function to import and use. If Rust supported defining a [default type](https://rust-lang.github.io/rfcs/0213-defaulted-type-params.html) we would not have this problem.

Yeah :/ Not too sure what the right compromise is in this case.

**Comment by Virgiel** [comment]
Good documentation and examples is all we can do I guess 🤷🏻‍♂️

**Comment by LouisGariepy** [comment]
Codegen is practically done! I didn't publish on the PR yet because it's late and I I need to sleep on it, but so far so good :smile: .

**Comment by LouisGariepy** [comment]
On track to be solved by #52.
